src/gadgets/tower_extension/algebraic_torus.rs

```diff
@@ -0,0 +1,299 @@
+        // TODO: w_inverse can be known in compile-time so there is no need to allocate it in run-time.
+        let mut w_inverse = w.inverse(cs);
+        let mut denominator = w.frobenius_map(cs, power);
+        let denominator = denominator.mul(cs, &mut w_inverse);
+
+        // Asserting that c1 is zero since denominator must be a pure Fq6 element.
+        let boolean_true = Boolean::allocated_constant(cs, true);
+        let c1_is_zero = g.c1.is_zero(cs);
+        Boolean::enforce_equal(cs, &c1_is_zero, &boolean_true);
+        let mut denominator = denominator.c0.clone();
+
+        // frob_map(g, i) / \gamma^{(p^i-1)/2}
+        let encoding = numerator.div(cs, &mut denominator);
+        Self::new(encoding)
+    }
+
+    /// Computes the product of two Torus elements using the formula
+    ///
+    /// `(g, g') -> (g * g' + \gamma) / (g + g')`
+    ///
+    /// The formula handles the exceptional case when `g + g'` is zero.
+    pub fn mul<CS, const SAFE: bool>(&mut self, cs: &mut CS, other: &mut Self) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let params = self.get_params();
+        let mut gamma = Fq6::gamma(cs, params);
+
+        // g1 <- g, g2 <- g'
+        let mut g1 = self.encoding.clone();
+        let mut g2 = other.encoding.clone();
+
+        // x <- g * g' + \gamma
+        let mut x = g1.mul(cs, &mut g2);
+        let mut x = x.add(cs, &mut gamma);
+
+        // y <- g + g'
+        let mut y = g1.add(cs, &mut g2);
+
+        let encoding = if SAFE {
+            // Exception occurs when g = -g'. To account for such case,
+            // the following formula is used (flag = (y == 0)? 1 : 0 --- exception case):
+            // result <- (x - flag * x) / (g + g' + flag)
+
+            let mut zero = Fq6::zero(cs, params);
+            let exception = y.equals(cs, &mut zero);
+            let mut flag = Fq6::from_boolean(cs, exception, params);
+
+            let mut numerator = flag.mul(cs, &mut x);
+            let mut numerator = x.sub(cs, &mut numerator);
+            let mut denominator = y.add(cs, &mut flag);
+            numerator.div(cs, &mut denominator)
+        } else {
+            // Here we do not check whether g = -g' since the function is unsafe
+            x.div(cs, &mut y)
+        };
+
+        Self::new(encoding)
+    }
+
+    pub fn pow_u32<CS, S: AsRef<[u64]>, const SAFE: bool>(
+        &mut self,
+        cs: &mut CS,
+        exponent: S,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let mut one = Fq12::one(cs, self.get_params());
+        let mut result = Self::compress::<CS, false>(cs, &mut one);
+        let mut found_one = false;
+
+        for i in BitIterator::new(exponent) {
+            if found_one {
+                result = result.square::<CS, SAFE>(cs);
+            } else {
+                found_one = i;
+            }
+
+            if i {
+                result = result.mul::<CS, SAFE>(cs, self);
+            }
```

Comment:
this is non-homogenous and needs to be updated to use conditional selection gates